src/data/list/join.lean

```diff
@@ -148,4 +148,23 @@ begin
   { simp },
 end
 
+/-- We can rebracket `x ++ (l₁ ++ x) ++ (l₂ ++ x) ++ ... ++ (lₙ ++ x)` to
+`(x ++ l₁) ++ (x ++ l₂) ++ ... ++ (x ++ lₙ) ++ x` where `L = [l₁, l₂, ..., lₙ]`. -/
+lemma append_join_append (L : list (list α)) (x : list α) :
+  x ++ (list.map (λ l, l ++ x) L).join = (list.map (λ l, x ++ l) L).join ++ x :=
+begin
+  induction L,
+  { rw [map_nil, join, append_nil, map_nil, join, nil_append] },
+  { rw [map_cons, join, map_cons, join, append_assoc, L_ih, append_assoc, append_assoc] },
+end
+
+/-- Reversing a list join is the same as reversing the order of parts and reversing all parts. -/
+lemma reverse_join (L : list (list α)) :
+  L.join.reverse = (list.map list.reverse L).reverse.join :=
```

Comment:
I think it would be also good to prove the similar lemma
```lean
  L.reverse.join = (list.map list.reverse L).join.reverse :=
```
which should follow from `reverse_reverse` and some cleanup.

Reply:
Done.